### PR TITLE
h264: replace usage of old BitReader with new

### LINKED ISSUE #43: h264: Replace usage of current BitReader with new BitReader.
Issue #30 needs to be addressed first. 

Usage of old BitReader that is defined in read.go should be replaced by new BitReader. We also need to fix calls to parsing functions that provide a nil BitReader.

### PR DESCRIPTION
**Pull request by saxon-milton**
It turned into a much bigger job than I could imagine. As you can see, it's mostly mechanical changes. 
Updates #43.

### CODE REVIEW

**Comment by kortschak** [comment]
Don't use the word "resolves" in front of an issue unless you want the issue closed on merging the mentioning PR. GitHub doesn't do NLP, so the partly there is not understood. Use "updates" instead (at the bottom of the opening PR comment).

**Comment by saxon-milton** [comment]
> Don't use the word "resolves" in front of an issue unless you want the issue closed on merging the mentioning PR. GitHub doesn't do NLP, so the partly there is not understood. Use "updates" instead (at the bottom of the opening PR comment).

Will do, cheers.

File: h264/bits/bitreader.go [hunk 1]
```diff
@@ -155,3 +157,14 @@ func (br *BitReader) PeekBits(n int) (uint64, error) {
 	r := (br.n >> uint(bits-n)) & ((1 << uint(n)) - 1)
 	return r, nil
 }
+
+// ByteAligned returns true if the reader position is at the start of a byte,
+// and false otherwise.
+func (br *BitReader) ByteAligned() bool {
```
Comment (kortschak) [inline h264/bits/bitreader.go:163]: There should be a test for this.

File: h264/bits/bitreader.go [hunk 2]
```diff
@@ -155,3 +157,14 @@ func (br *BitReader) PeekBits(n int) (uint64, error) {
...
+	return br.bits == 0
+}
+
+// BytesRead returns the number of bytes that have been read by the BitReader.
+func (br *BitReader) BytesRead() int {
```
Comment (kortschak) [inline h264/bits/bitreader.go:168]: This doesn't seem necessary. You can either query the underlying reader or just do the accounting by the caller.

File: h264/cabac.go [hunk 3]
```diff
@@ -1,6 +1,9 @@
 package h264
 
-import "github.com/pkg/errors"
+import (
+	"github.com/ausocean/h264decode/h264/bits"
+	"github.com/pkg/errors"
```
Comment (kortschak) [inline h264/cabac.go:5]: See previous comments about introducing this as a dep. Note that mrmod's code does not import it.

**Review by kortschak — COMMENTED** [review]
This doesn't look mostly mechanical. Is there a suggested review scheme?

File: h264/bits/bitreader.go (hunk 2 above)
Comment (saxon-milton) [reply h264/bits/bitreader.go:168]: So the caller should query the underlying reader for this information ? Is there an io.Reader implementation that provides this information ?

**Comment by saxon-milton** [comment]
> This doesn't look mostly mechanical. Is there a suggested review scheme?

Maybe I have a different (or incorrect) definition of mechanical. A lot of the changes in this PR are of this form,
`<some struct field> = b.NextField(<string name of field>,numberOfBits)`
to 
`b, err = br.ReadBits(numberOfBits)`
`if err != nil {`
`    return errors.Wrap(err,"could not read some field")`
`}`
`<some struct field> = int(b)`

Also, I've removed the old BitReader, and changed some of the old BitReaders methods to not be methods for a particular struct.

File: h264/bits/bitreader.go (hunk 2 above)
Comment (saxon-milton) [reply h264/bits/bitreader.go:168]: I think this method is stupid too, but I'm just trying not to break the current code. If I get rid of this, and we do the querying by the caller, there's going to be a lot more code added to this PR.

Comment (kortschak) [reply h264/bits/bitreader.go:168]: OK. Leave it as is. Why do we need to know the number of bytes that were read?

**Pull request merged**